File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/integrations/base/ssh/SshBastionContainer.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.base.ssh;

import static io.airbyte.cdk.integrations.base.ssh.SshTunnel.TunnelMethod.SSH_KEY_AUTH;
import static io.airbyte.cdk.integrations.base.ssh.SshTunnel.TunnelMethod.SSH_PASSWORD_AUTH;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.collect.ImmutableMap;
import io.airbyte.cdk.integrations.util.HostPortResolver;
import io.airbyte.commons.json.Jsons;
import java.io.IOException;
import java.util.List;
import java.util.Objects;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.testcontainers.containers.Container;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.JdbcDatabaseContainer;
import org.testcontainers.containers.Network;
import org.testcontainers.images.builder.ImageFromDockerfile;

public class SshBastionContainer implements AutoCloseable {

  private static final String SSH_USER = "sshuser";
  private static final String SSH_PASSWORD = "secret";
  private GenericContainer bastion;

  public void initAndStartBastion(final Network network) {
    bastion = new GenericContainer(
        new ImageFromDockerfile("bastion-test")
            .withFileFromClasspath("Dockerfile", "bastion/Dockerfile"))
                .withNetwork(network)
                .withExposedPorts(22);
    bastion.start();
  }

  public JsonNode getTunnelMethod(final SshTunnel.TunnelMethod tunnelMethod,
                                  final boolean innerAddress)
      throws IOException, InterruptedException {
    final var containerAddress = innerAddress ? getInnerContainerAddress(bastion) : getOuterContainerAddress(bastion);
    return Jsons.jsonNode(ImmutableMap.builder()
        .put("tunnel_host",
            Objects.requireNonNull(containerAddress.left))
        .put("tunnel_method", tunnelMethod)
        .put("tunnel_port", containerAddress.right)
        .put("tunnel_user", SSH_USER)
        .put("tunnel_user_password", tunnelMethod.equals(SSH_PASSWORD_AUTH) ? SSH_PASSWORD : "")
        .put("ssh_key", tunnelMethod.equals(SSH_KEY_AUTH) ? bastion.execInContainer("cat", "var/bastion/id_rsa").getStdout() : "")
        .build());
  }

  public JsonNode getTunnelConfig(final SshTunnel.TunnelMethod tunnelMethod,
                                  final ImmutableMap.Builder<Object, Object> builderWithSchema,
                                  final boolean innerAddress)
      throws IOException, InterruptedException {
    return Jsons.jsonNode(builderWithSchema
        .put("tunnel_method", getTunnelMethod(tunnelMethod, innerAddress))
        .build());
  }

  public ImmutableMap.Builder<Object, Object> getBasicDbConfigBuider(final JdbcDatabaseContainer<?> db) {
    return getBasicDbConfigBuider(db, db.getDatabaseName());
  }

  public ImmutableMap.Builder<Object, Object> getBasicDbConfigBuider(final JdbcDatabaseContainer<?> db, final List<String> schemas) {
    return getBasicDbConfigBuider(db, db.getDatabaseName()).put("schemas", schemas);
  }

  public ImmutableMap.Builder<Object, Object> getBasicDbConfigBuider(final JdbcDatabaseContainer<?> db, final String schemaName) {
    return ImmutableMap.builder()
        .put("host", Objects.requireNonNull(HostPortResolver.resolveHost(db)))
        .put("username", db.getUsername())
        .put("password", db.getPassword())
        .put("port", HostPortResolver.resolvePort(db))
        .put("database", schemaName)
        .put("ssl", false);
  }

  public void stopAndCloseContainers(final JdbcDatabaseContainer<?> db) {
    bastion.stop();
    bastion.close();
    db.stop();
    db.close();
  }

  public void stopAndClose() {
    bastion.close();
  }

  @Override
  public void close() {
    stopAndClose();
  }

  public GenericContainer getContainer() {
    return bastion;
  }

  /**
   * Returns the inner docker network ip address and port of a container. This can be used to reach a
   * container from another container running on the same network
   *
   * @param container container
   * @return a pair of host and port
   */
  public static ImmutablePair<String, Integer> getInnerContainerAddress(final Container container) {
    return ImmutablePair.of(
        container.getContainerInfo().getNetworkSettings().getNetworks().entrySet().stream().findFirst().get().getValue().getIpAddress(),
        (Integer) container.getExposedPorts().stream().findFirst().get());
  }

  /**
   * Returns the outer docker network ip address and port of a container. This can be used to reach a
   * container from the host machine
   *
   * @param container container
   * @return a pair of host and port
   */
  public static ImmutablePair<String, Integer> getOuterContainerAddress(final Container container) {
    return ImmutablePair.of(
        container.getHost(),
        container.getFirstMappedPort());
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/kotlin/io/airbyte/cdk/integrations/base/ssh/SshBastionContainer.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.base.ssh

import com.fasterxml.jackson.databind.JsonNode
import com.google.common.collect.ImmutableMap
import io.airbyte.cdk.integrations.util.HostPortResolver
import io.airbyte.cdk.testutils.ContainerFactory
import io.airbyte.commons.json.Jsons
import java.io.IOException
import java.util.*
import java.util.function.Consumer
import org.apache.commons.lang3.tuple.ImmutablePair
import org.testcontainers.containers.Container
import org.testcontainers.containers.GenericContainer
import org.testcontainers.containers.JdbcDatabaseContainer
import org.testcontainers.containers.Network
import org.testcontainers.images.builder.ImageFromDockerfile
import org.testcontainers.utility.DockerImageName

class SshBastionContainer : AutoCloseable {
    class SshBastionContainerFactory : ContainerFactory<GenericContainer<*>>() {
        override fun createNewContainer(imageName: DockerImageName): GenericContainer<*> {
            val container: GenericContainer<*> =
                GenericContainer<Nothing>(
                        ImageFromDockerfile("bastion-test")
                            .withFileFromClasspath("Dockerfile", "bastion/Dockerfile")
                    )
                    .withExposedPorts(22)
            return container
        }

        fun exclusive(network: Network): GenericContainer<*> {
            val imageModifier = Consumer { c: GenericContainer<*> -> c.withNetwork(network) }
            val container =
                super.exclusive(
                    "bastion-test",
                    NamedContainerModifierImpl("withNetwork", imageModifier)
                )
            return container
        }
    }

    var container: GenericContainer<*>? = null
        private set

    fun initAndStartBastion(network: Network) {
        container = factory!!.exclusive(network)
        container!!.start()
    }

    @Throws(IOException::class, InterruptedException::class)
    fun getTunnelMethod(tunnelMethod: SshTunnel.TunnelMethod, innerAddress: Boolean): JsonNode? {
        val containerAddress =
            if (innerAddress) getInnerContainerAddress(container!!)
            else getOuterContainerAddress(container!!)
        return Jsons.jsonNode(
            ImmutableMap.builder<Any?, Any?>()
                .put("tunnel_host", Objects.requireNonNull(containerAddress!!.left))
                .put("tunnel_method", tunnelMethod)
                .put("tunnel_port", containerAddress.right)
                .put("tunnel_user", SSH_USER)
                .put(
                    "tunnel_user_password",
                    if (tunnelMethod == SshTunnel.TunnelMethod.SSH_PASSWORD_AUTH) SSH_PASSWORD
                    else ""
                )
                .put(
                    "ssh_key",
                    if (tunnelMethod == SshTunnel.TunnelMethod.SSH_KEY_AUTH)
                        container!!.execInContainer("cat", "var/bastion/id_rsa").stdout
                    else ""
                )
                .build()
        )
    }

    @Throws(IOException::class, InterruptedException::class)
    fun getTunnelConfig(
        tunnelMethod: SshTunnel.TunnelMethod,
        builderWithSchema: ImmutableMap.Builder<Any, Any>,
        innerAddress: Boolean
    ): JsonNode? {
        return Jsons.jsonNode(
            builderWithSchema
                .put("tunnel_method", getTunnelMethod(tunnelMethod, innerAddress))
                .build()
        )
    }

    fun getBasicDbConfigBuider(db: JdbcDatabaseContainer<*>): ImmutableMap.Builder<Any, Any> {
        return getBasicDbConfigBuider(db, db.databaseName)
    }

    fun getBasicDbConfigBuider(
        db: JdbcDatabaseContainer<*>,
        schemas: MutableList<String>
    ): ImmutableMap.Builder<Any, Any> {
        return getBasicDbConfigBuider(db, db.databaseName).put("schemas", schemas)
    }

    fun getBasicDbConfigBuider(
        db: JdbcDatabaseContainer<*>,
        schemaName: String
    ): ImmutableMap.Builder<Any, Any> {
        return ImmutableMap.builder<Any, Any>()
            .put("host", Objects.requireNonNull(HostPortResolver.resolveHost(db)))
            .put("username", db.username)
            .put("password", db.password)
            .put("port", HostPortResolver.resolvePort(db))
            .put("database", schemaName)
            .put("ssl", false)
    }

    fun stopAndCloseContainers(db: JdbcDatabaseContainer<*>) {
        container!!.stop()
        container!!.close()
        db.stop()
        db.close()
    }

    fun stopAndClose() {
        container!!.close()
    }

    override fun close() {
        stopAndClose()
    }

    companion object {
        private val factory: SshBastionContainerFactory? = SshBastionContainerFactory()

        private val SSH_USER: String = "sshuser"
        private val SSH_PASSWORD: String = "secret"

        @JvmStatic
        /**
         * Returns the inner docker network ip address and port of a container. This can be used to
         * reach a container from another container running on the same network
         *
         * @param container container
         * @return a pair of host and port
         */
        fun getInnerContainerAddress(container: Container<*>): ImmutablePair<String, Int> {
            return ImmutablePair.of(
                container.containerInfo.networkSettings.networks.entries.first().value.ipAddress,
                container.exposedPorts.first()
            )
        }

        @JvmStatic
        /**
         * Returns the outer docker network ip address and port of a container. This can be used to
         * reach a container from the host machine
         *
         * @param container container
         * @return a pair of host and port
         */
        fun getOuterContainerAddress(container: Container<*>): ImmutablePair<String, Int> {
            return ImmutablePair.of(container.host, container.firstMappedPort)
        }
    }
}
